File: frontend/src/types/email.ts
export interface EmailConfig {
  id: number
  user_id: number
  email_address: string
  imap_server: string
  imap_port: number
  username: string
  scan_days: number
  is_active: boolean
  last_scan_time?: string
  created_at: string
  updated_at: string
}

export interface EmailConfigCreate {
  email_address: string
  imap_server: string
  imap_port: number
  username: string
  password: string
  scan_days?: number
}

export interface EmailConfigUpdate {
  email_address?: string
  imap_server?: string
  imap_port?: number
  username?: string
  password?: string
  scan_days?: number
  is_active?: boolean
}

export interface EmailConnectionTest {
  email_address: string
  imap_server: string
  imap_port: number
  username: string
  password: string
}

export interface ManualScanRequest {
  config_id?: number
  days?: number
  force?: boolean
}

export interface ScanTaskStatus {
  state: string
  current?: number
  total?: number
  status?: string
  result?: any
  error?: string
}

export interface EmailScanResult {
  config_id: number
  email_address: string
  processed_count: number
  success_count: number
  error_count: number
  errors: string[]
  processed_files: Array<{
    filename: string
    type: string
    status: string
    invoice_id?: string
  }>
}

// 邮件列表相关类型
export interface Email {
  id: string
  user_id: number
  message_id: string
  subject?: string
  sender?: string
  recipient?: string
  date_sent?: string
  date_received?: string
  body_text?: string
  body_html?: string
  has_attachments: boolean
  attachment_count: number
  attachment_info?: Array<{
    filename: string
    content_type: string
    size: number
    is_pdf: boolean
  }>
  invoice_scan_status: string
  invoice_count: number
  scan_result?: Record<string, any>
  processing_status: string
  error_message?: string
  created_at: string
  updated_at: string
  scanned_at?: string
}

export interface EmailFilter {
  invoice_scan_status?: string
  processing_status?: string
  sender?: string
  subject?: string
  has_attachments?: boolean
  has_invoice?: boolean
  date_from?: string
  date_to?: string
}

export interface EmailListResponse {
  emails: Email[]
  total: number
  page: number
  size: number
  pages: number
}

export interface EmailStatistics {
  period_days: number
  total_emails: number
  emails_with_attachments: number
  emails_with_invoices: number
  emails_without_invoices: number
  total_invoices_found: number
  scan_status_breakdown: Record<string, number>
  processing_status_breakdown: Record<string, number>
  attachment_rate: number
  invoice_detection_rate: number
}

export interface EmailBatchOperation {
  email_ids: string[]
  operation: 'rescan' | 'delete'
}

export interface EmailBatchOperationResponse {
  success: boolean
  total: number
  processed: number
  failed: number
  failed_details?: Array<{
    email_id: string
    reason: string
  }>
  message: string
}

export interface EmailUpdate {
  subject?: string
  invoice_scan_status?: 'pending' | 'no_invoice' | 'has_invoice'
  processing_status?: string
  invoice_count?: number
  scan_result?: Record<string, any>
  error_message?: string
}

File: backend/app/schemas/email.py
"""
邮件相关的Pydantic模型
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


class EmailBase(BaseModel):
    """邮件基础模型"""
    subject: Optional[str] = None
    sender: Optional[str] = None
    recipient: Optional[str] = None
    date_sent: Optional[datetime] = None
    date_received: Optional[datetime] = None
    body_text: Optional[str] = None
    body_html: Optional[str] = None
    has_attachments: bool = False
    attachment_count: int = 0
    attachment_info: Optional[List[Dict[str, Any]]] = None


class EmailCreate(EmailBase):
    """创建邮件模型"""
    message_id: str = Field(..., description="邮件唯一标识")


class EmailUpdate(BaseModel):
    """更新邮件模型"""
    subject: Optional[str] = None
    invoice_scan_status: Optional[str] = None
    processing_status: Optional[str] = None
    invoice_count: Optional[int] = None
    scan_result: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None


class Email(EmailBase):
    """邮件响应模型"""
    id: str
    user_id: int
    message_id: str
    invoice_scan_status: str
    invoice_count: int
    scan_result: Optional[Dict[str, Any]] = None
    processing_status: str
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    scanned_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class EmailFilter(BaseModel):
    """邮件筛选模型"""
    invoice_scan_status: Optional[str] = Field(None, description="发票扫描状态")
    processing_status: Optional[str] = Field(None, description="处理状态")
    sender: Optional[str] = Field(None, description="发送者")
    subject: Optional[str] = Field(None, description="主题关键词")
    has_attachments: Optional[bool] = Field(None, description="是否有附件")
    has_invoice: Optional[bool] = Field(None, description="是否检测到发票")
    date_from: Optional[datetime] = Field(None, description="开始日期")
    date_to: Optional[datetime] = Field(None, description="结束日期")


class PaginationParams(BaseModel):
    """分页参数"""
    page: int = Field(1, ge=1, description="页码")
    size: int = Field(20, ge=1, le=100, description="每页大小")


class EmailListResponse(BaseModel):
    """邮件列表响应模型"""
    emails: List[Email]
    total: int
    page: int
    size: int
    pages: int

    @classmethod
    def create(cls, emails: List[Email], total: int, page: int, size: int):
        pages = (total + size - 1) // size
        return cls(
            emails=emails,
            total=total,
            page=page,
            size=size,
            pages=pages
        )


class EmailStatistics(BaseModel):
    """邮件统计模型"""
    period_days: int
    total_emails: int
    emails_with_attachments: int
    emails_with_invoices: int
    emails_without_invoices: int
    total_invoices_found: int
    scan_status_breakdown: Dict[str, int]
    processing_status_breakdown: Dict[str, int]
    attachment_rate: float
    invoice_detection_rate: float


class EmailBatchOperation(BaseModel):
    """邮件批量操作模型"""
    email_ids: List[str] = Field(..., description="邮件ID列表")
    operation: str = Field(..., description="操作类型: rescan, delete")


class EmailBatchOperationResponse(BaseModel):
    """邮件批量操作响应模型"""
    success: bool
    total: int
    processed: int
    failed: int
    failed_details: Optional[List[Dict[str, Any]]] = None
    message: str